### PR TITLE
BART with non-gaussian likelihoods

### PR DESCRIPTION
**Pull request by aloctavodia**
This allows to use other likelihoods than the Gaussian one. Adds a new argument `inv_link` to allow the user to specify the proper inverse link function, defaults to identity.  This also introduce an `scale` argument that allows to control the proposal of the leaf values. For the Gaussian likelihoods (and probably other unbounded likelihoods like the Student T) the default scale is ok, for other model like the binomial it may need to be adjusted for better results. It seems LOO can be use for this purpose. 

A simple example

```python
from scipy.special import expit

with pm.Model() as model:
    μ = pm.BART('μ', X, Y, m=50, scale=0.25, inv_link=expit)
    y = pm.Bernoulli('y', μ, observed=Y)
    trace = pm.sample(1000, return_inferencedata=True)
```

test_01b

### CODE REVIEW

**Comment by junpenglao** [comment]
Add a small test?

**Review by junpenglao — APPROVED** [review]
LGTM

File: pymc3/distributions/bart.py [hunk 1]
```diff
@@ -257,6 +278,24 @@ def rvs(self):
                     return i
 
 
+class NormalSampler:
+    def __init__(self, sigma):
+        self.size = 5000
```
Comment (canyon289) [inline pymc3/distributions/bart.py:283]: Why 5000?

File: pymc3/step_methods/pgbart.py [hunk 2]
```diff
@@ -86,12 +86,13 @@ def __init__(self, vars=None, num_particles=10, max_stages=5000, chunk="auto", m
...
     def astep(self, _):
         bart = self.bart
+        inv_link = bart.inv_link
         num_observations = bart.num_observations
         variable_inclusion = np.zeros(bart.num_variates, dtype="int")
 
         # For the tunning phase we restrict max_stages to a low number, otherwise it is almost sure
         # we will reach max_stages given that our first set of m trees is not good at all.
-        # Can set max_stages as a function of the number of variables/dimensions?
+        # Can set max_stages as a function of the number of variables/dimensions? XXX
```
Comment (canyon289) [inline pymc3/step_methods/pgbart.py:95]: Whats XXX?

File: pymc3/tests/test_bart.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+    idx_data_points = np.array(range(bart.num_observations), dtype="int32")
+    available_splitting_rules = bart.get_available_splitting_rules(
+        idx_data_points, idx_split_variable
+    )
+    assert available_splitting_rules.size == 1
+    np.testing.assert_almost_equal(available_splitting_rules, np.array([-3.3]), 1)
+
+
+def test_model():
+    np.random.seed(212480)
```
Comment (canyon289) [inline pymc3/tests/test_bart.py:94]: Im not sure about this but should random seeds be generated with numpys rng thing?

**Comment by canyon289** [comment]
Small comments, otherwise im good

**Review by canyon289 — APPROVED** [review]

File: pymc3/step_methods/pgbart.py (hunk 2 above)
Comment (aloctavodia) [reply pymc3/step_methods/pgbart.py:95]: A note to myself to check this later, I introduced the restriction during the early development. Now it seems it is not necessary anymore , but I did not have the time to tested thoroughly.

File: pymc3/distributions/bart.py (hunk 1 above)
Comment (aloctavodia) [reply pymc3/distributions/bart.py:283]: larger cache size does not seem to provide any speed up.

File: pymc3/tests/test_bart.py (hunk 3 above)
Comment (aloctavodia) [reply pymc3/tests/test_bart.py:94]: Also not sue, i think this is something we do in many of our tests, so I will defer until we decide on an overall action.

**Pull request merged**